Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+
+#include "config.h"
+#include "TextIndicatorStyleController.h"
+
+#include "TextIndicatorStyle.h"
+#include "UnifiedTextReplacementController.h"
+#include "WebPage.h"
+#include <WebCore/DocumentMarkerController.h>
+#include <WebCore/FocusController.h>
+#include <WebCore/RenderedDocumentMarker.h>
+#include <WebCore/SimpleRange.h>
+#include <WebCore/TextIndicator.h>
+#include <WebCore/TextIterator.h>
+
+namespace WebKit {
+
+TextIndicatorStyleController::TextIndicatorStyleController(WebPage& webPage)
+    : m_webPage(webPage)
+{
+}
+
+RefPtr<WebCore::Document> TextIndicatorStyleController::document() const
+{
+    if (!m_webPage) {
+        ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+
+    RefPtr corePage = m_webPage->corePage();
+    if (!corePage) {
+        ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+
+    RefPtr frame = corePage->checkedFocusController()->focusedOrMainFrame();
+    if (!frame) {
+        ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+
+    return frame->document();
+}
+
+std::optional<WebCore::SimpleRange> TextIndicatorStyleController::contextRangeForTextIndicatorStyle(const WTF::UUID& uuid) const
+{
+    auto sessionRange = m_webPage->unifiedTextReplacementController().contextRangeForSessionWithUUID(uuid);
```

Comment:
I don't think this variable is used anywhere other than the next two lines? Can move the assignment into the if.